draft-cpbs-pseudorandom-ctls.md

```diff
@@ -172,6 +171,29 @@ CTLSPlaintext records are subject to an additional decipherment step:
 2. Let `tweak` be `"client datagram hs" + profile_id + Handshake.msg_type` if sent by the client, or `"server datagram hs" + profile_id + Handshake.msg_type` if sent by the server.
 3. Replace `Handshake.body` with `STPRP-Decipher(key, tweak, Handshake.body)`.
 
+## Optional defense against protocol confusion {#confusion-defense}
+
+The procedure described above is sufficient to render the bitstream pseudorandom to a third party when both peers are operating correctly.  However, if a malicious client or server can coerce its peer into sending particular plaintext (as is common in web browsers), it can choose plaintext with knowledge of the encryption keys, in order to produce ciphertext that has visible structure to a third party.  This technique can be used to mount protocol confusion attacks {{SLIPSTREAM}}.
+
+This attack is particularly straightforward when using the AES-GCM or ChaCha20-Poly1305 cipher suites, as much of the ciphertext is encrypted by XOR with a stream cipher.  A malicious peer in this threat model can choose desired ciphertext, XOR it with the keystream to produce the malicious plaintext, and rely on the other peer's encryption stage to reverse the encryption and reveal the desired ciphertext.
+
+As a defense for this threat model, the Pseudorandom cTLS extension supports two optional keys named "client-recipher" and "server-recipher".  Each key's value is an integer `E` in the range 0..16 indicating how much entropy to add.  When the "client-recipher" key is present, the client MUST modify each outgoing ciphertext message as follows:
+
+1. Let `tweak = "client " + (dtls ? "datagram " : "") + "body"
+2. Append the 64-bit Sequence Number to `tweak`.
+3. Let `R` be a string containing `E` fresh random bytes.
+4. Replace `CTLSCiphertext.encrypted_record` with `R + STPRP-Encipher(key, tweak + R, CTLSCiphertext.encrypted_record)`.
+
+The server MUST apply a similar transformation if the "server-recipher" key is present.
+
+This transformation does not alter the `Length` field in the Unified Header, so it does not reduce the maximum plaintext record size.  However, it does increase the output message size, which may impact MTU calculations in DTLS.
+
+If `R` is computed using a pseudo-random number generator, it MUST be cryptographically secure and keyed with at least 16 bytes of entropy that is not known to the peer.
+
+When `E=0`, this process fails to defend against ciphertext confusion attacks for a general STPRP and AEAD.  A malicious peer can easily construct plaintext such that the STPRP output contains a desired bytestring.  Templates MUST NOT specify `E=0` unless the STPRP is known to be safe from known-key distinguishing attacks {{?KNOWNKEY=DOI.10.1007/978-3-662-43933-3_18}}, which most are not.
```

Comment:
I suspect that resistance against known-key distinguishing attacks isn't going to be sufficient because the attack [here](https://github.com/bemasc/pseudorandom-ctls/pull/10#pullrequestreview-824143503) is effective even if the STPRP is an ideal cipher. In fact, it may turn out that this security property is necessary *even when `E=16`*.

Reply:
I'm not sure what you mean exactly, but I think the mental gymnastics have gotten too convoluted here, so I've removed this whole paragraph.  If a TSPRP with the right properties magically appears in the future, we can have this discussion then.

Reply:
That works for me.

Reply:
The tricky thing here is that the properties you want from the TSPRP are going to depend on the strength of the attack you want to defend against. I'll try to articulate this in an issue.